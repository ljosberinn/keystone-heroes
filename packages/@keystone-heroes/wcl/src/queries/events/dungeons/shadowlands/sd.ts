import type { ApplyBuffEvent, BeginCastEvent } from "../../types";
import { createIsSpecificEvent } from "../../utils";

export const SD_LANTERN_OPENING = 340_013;
export const SD_LANTERN_BUFF = 340_433;

/**
 * see https://www.warcraftlogs.com/reports/mQn47kFCbpThBtrG#fight=15&type=summary&view=events&pins=2%24Off%24%23909049%24expression%24type%20%3D%20%22applybuff%22%20and%20ability.id%20%3D%20340433%5E2%24Off%24%23909049%24expression%24ability.id%20%3D%20340013%20and%20type%20%3D%22begincast%22
 */
const lanternOpeningExpression = `type ="begincast" and ability.id = ${SD_LANTERN_OPENING}`;
const lanternBuffExpression = `type="applybuff" and ability.id = ${SD_LANTERN_BUFF}`;

export const filterExpression = [
  lanternOpeningExpression,
  lanternBuffExpression,
];

export const isSdLanternOpeningEvent = createIsSpecificEvent<BeginCastEvent>({
  type: "begincast",
  abilityGameID: SD_LANTERN_OPENING,
});

export const isSdLanternBuffEvent = createIsSpecificEvent<ApplyBuffEvent>({
  type: "applybuff",
  abilityGameID: SD_LANTERN_BUFF,
});
