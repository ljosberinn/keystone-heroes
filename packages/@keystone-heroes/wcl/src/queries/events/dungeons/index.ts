import { DungeonIDs } from "@keystone-heroes/db/data";
import type { PlayableClass } from "@keystone-heroes/db/types";

import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  DeathEvent,
  HealEvent,
  RemoveBuffEvent,
} from "../../events/types";
import { filterExpression as dosFilterExpression, getDOSEvents } from "./dos";
import { filterExpression as hoaFilterExpression, getHOAEvents } from "./hoa";
import { filterExpression as nwFilterExpression, getNWEvents } from "./nw";
import { filterExpression as pfFilterExpression, getPFEvents } from "./pf";
import { filterExpression as sdFilterExpression, getSDEvents } from "./sd";
import { filterExpression as soaFilterExpression, getSOAEvents } from "./soa";
import { filterExpression as topFilterExpression, getTOPEvents } from "./top";

type DungeonWithEvents = Exclude<DungeonIDs, DungeonIDs.MISTS_OF_TIRNA_SCITHE>;

const isDungeonWithEvent = (id: DungeonIDs): id is DungeonWithEvents =>
  id in dungeonExpressionMap;

const dungeonExpressionMap: Record<DungeonWithEvents, string[]> = {
  [DungeonIDs.DE_OTHER_SIDE]: dosFilterExpression,
  [DungeonIDs.HALLS_OF_ATONEMENT]: hoaFilterExpression,
  [DungeonIDs.PLAGUEFALL]: pfFilterExpression,
  [DungeonIDs.SANGUINE_DEPTHS]: sdFilterExpression,
  [DungeonIDs.SPIRES_OF_ASCENSION]: soaFilterExpression,
  [DungeonIDs.THEATRE_OF_PAIN]: topFilterExpression,
  [DungeonIDs.THE_NECROTIC_WAKE]: nwFilterExpression,
};

export const getDungeonExpression = (id: DungeonIDs): string[] =>
  isDungeonWithEvent(id) ? dungeonExpressionMap[id] : [];

export const filterDungeonEvents = (
  allEvents: AllTrackedEventTypes,
  dungeonID: DungeonIDs,
  playerMetaInformation: { actorID: number; class: PlayableClass }[]
):
  | ApplyDebuffEvent[]
  | CastEvent[]
  | (DeathEvent | DamageEvent | ApplyBuffEvent)[]
  | (BeginCastEvent | ApplyBuffEvent | RemoveBuffEvent | ApplyBuffStackEvent)[]
  | (DamageEvent | HealEvent)[] => {
  switch (dungeonID) {
    case DungeonIDs.DE_OTHER_SIDE:
      return getDOSEvents(allEvents);
    case DungeonIDs.HALLS_OF_ATONEMENT:
      return getHOAEvents(allEvents);
    case DungeonIDs.PLAGUEFALL:
      return getPFEvents(allEvents, playerMetaInformation);
    case DungeonIDs.SANGUINE_DEPTHS:
      return getSDEvents(allEvents);
    case DungeonIDs.SPIRES_OF_ASCENSION:
      return getSOAEvents(allEvents);
    case DungeonIDs.THEATRE_OF_PAIN:
      return getTOPEvents(allEvents);
    case DungeonIDs.THE_NECROTIC_WAKE:
      return getNWEvents(allEvents);
    default:
      return [];
  }
};
