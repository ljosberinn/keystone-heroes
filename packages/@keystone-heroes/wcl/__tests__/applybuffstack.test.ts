import { SD_LANTERN_BUFF } from "../src/queries/events/dungeons/sd";
import type { ApplyBuffStackEvent } from "../src/queries/events/types";
import { applyBuffStackProcessor } from "../src/transform/events/applybuffstack";
import allEvents from "./fixtures/allEvents.json";

describe("applybuffstack", () => {
  const params = {
    sourcePlayerID: null,
    targetPlayerID: 1,
    sourceNPCID: null,
    targetNPCID: null,
  };

  test("works for SD Lantern Buff", () => {
    const result = allEvents
      .filter(
        (event): event is ApplyBuffStackEvent =>
          event.type === "applybuffstack" &&
          event.abilityGameID === SD_LANTERN_BUFF
      )
      .map((event) => applyBuffStackProcessor(event, params));

    expect(result).toMatchSnapshot();
  });

  test("returns null for unknown applybuffstack events", () => {
    expect(
      applyBuffStackProcessor(
        {
          type: "applybuffstack",
          abilityGameID: 1,
          sourceID: 1,
          stack: 0,
          targetID: 0,
          timestamp: 1,
        },
        params
      )
    ).toBeNull();
  });

  test("returns null for applybuffstack events without targetPlayerID", () => {
    expect(
      applyBuffStackProcessor(
        {
          type: "applybuffstack",
          abilityGameID: 1,
          sourceID: 1,
          stack: 0,
          targetID: 0,
          timestamp: 1,
        },
        {
          ...params,
          targetPlayerID: null,
        }
      )
    ).toBeNull();
  });
});
