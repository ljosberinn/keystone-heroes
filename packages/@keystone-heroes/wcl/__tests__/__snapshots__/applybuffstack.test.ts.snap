// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applybuffstack works for SD Lantern Buff 1`] = `
Array [
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 11578032,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 11578032,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 11578032,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 11578032,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 11578032,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 11578435,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 11578435,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 11578435,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 11578435,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 11578435,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 11578528,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 11578528,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 11578528,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 11578528,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 11578528,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 11583194,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 11583194,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 11583194,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 11583194,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 11583194,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 11585639,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 11585639,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 11585639,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 11585639,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 11585639,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 11590332,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 11590332,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 11590332,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 11590332,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 11590332,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 11594225,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 11594225,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 11594225,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 11594225,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 11594225,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 11594335,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 11594335,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 11594335,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 11594335,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 11594335,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 10,
    "targetPlayerID": 1,
    "timestamp": 11596689,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 10,
    "targetPlayerID": 1,
    "timestamp": 11596689,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 10,
    "targetPlayerID": 1,
    "timestamp": 11596689,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 10,
    "targetPlayerID": 1,
    "timestamp": 11596689,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 10,
    "targetPlayerID": 1,
    "timestamp": 11596689,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 12270637,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 12270637,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 12270637,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 12270637,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 12270637,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 12270637,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 12270637,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 12270637,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 12270637,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 12270637,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 12271043,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 12271043,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 12271043,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 12271043,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 12271043,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 12271214,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 12271214,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 12271214,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 12271214,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 12271214,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 12277302,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 12277302,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 12277302,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 12277302,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 12277302,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 12279309,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 12279309,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 12279309,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 12279309,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 12279309,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 12287914,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 12287914,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 12287914,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 12287914,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 12287914,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 12291261,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 12291261,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 12291261,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 12291278,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 12291278,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 12640578,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 12640578,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 12640578,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 12640578,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 12640578,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 12641866,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 12641866,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 12641866,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 12641866,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 12641866,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 12653658,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 12653658,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 12653658,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 12653658,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 12653658,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 12653658,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 12653658,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 12653658,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 12653658,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 12653658,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 12655096,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 12655096,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 12655096,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 12655096,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 12655096,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 13483430,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 13483430,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 13483430,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 13483430,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 2,
    "targetPlayerID": 1,
    "timestamp": 13483430,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 13483981,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 13483981,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 13483981,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 13483981,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 3,
    "targetPlayerID": 1,
    "timestamp": 13483981,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 13484475,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 13484475,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 13484475,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 13484475,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 4,
    "targetPlayerID": 1,
    "timestamp": 13484475,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 13484569,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 13484569,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 13484569,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 13484569,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 5,
    "targetPlayerID": 1,
    "timestamp": 13484569,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 13485163,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 13485163,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 13485163,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 13485163,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 6,
    "targetPlayerID": 1,
    "timestamp": 13485163,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 13485163,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 13485163,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 13485163,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 13485163,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 7,
    "targetPlayerID": 1,
    "timestamp": 13485163,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 13485246,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 13485246,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 13485246,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 13485246,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 8,
    "targetPlayerID": 1,
    "timestamp": 13485246,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 13485565,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 13485565,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 13485565,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 13485565,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 9,
    "targetPlayerID": 1,
    "timestamp": 13485565,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 10,
    "targetPlayerID": 1,
    "timestamp": 13487006,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 10,
    "targetPlayerID": 1,
    "timestamp": 13487006,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 10,
    "targetPlayerID": 1,
    "timestamp": 13487006,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 10,
    "targetPlayerID": 1,
    "timestamp": 13487006,
  },
  Object {
    "abilityID": 340433,
    "eventType": "ApplyBuffStack",
    "stacks": 10,
    "targetPlayerID": 1,
    "timestamp": 13487029,
  },
]
`;
